atef/bin/config.py

```diff
@@ -0,0 +1,1528 @@
+
+    Consists of a tree visualization on the left that can be selected through
+    to choose which part of the tree to edit in the widget space on the right.
+
+    Parameters
+    ----------
+    config_file : ConfigurationFile
+        The config file object to use to build the tree.
+    """
+    filename = 'config_tree.ui'
+
+    bridge: QDataclassBridge
+    tree_widget: QTreeWidget
+
+    full_path: str
+
+    def __init__(
+        self,
+        *args,
+        config_file: ConfigurationFile,
+        full_path: Optional[str] = None,
+        **kwargs
+    ):
+        super().__init__(*args, **kwargs)
+        self.bridge = QDataclassBridge(config_file, parent=self)
+        self.full_path = full_path
+        self.last_selection: Optional[AtefItem] = None
+        self.built_widgets = set()
+        self.assemble_tree()
+        self.show_selected_display(self.overview_item)
+        self.tree_widget.itemPressed.connect(self.show_selected_display)
+
+    def debug_show_data(self, *args, **kwargs):
+        print(self.bridge.data)
+
+    def assemble_tree(self):
+        """
+        On startup, create the full tree.
+
+        TODO: properly fill from the config_file, currently creates only the
+        base tree with a single "overview" item.
+        """
+        self.tree_widget.setColumnCount(2)
+        self.tree_widget.setHeaderLabels(['Node', 'Type'])
+        self.overview_item = AtefItem(
+            widget_class=Overview,
+            widget_args=[self.bridge.configs, self.tree_widget],
+            name='Overview',
+            func_name='overview'
+        )
+        self.tree_widget.addTopLevelItem(self.overview_item)
+
+    def show_selected_display(self, item: AtefItem, *args, **kwargs):
+        """
+        Show the proper widget on the right when a tree row is selected.
+
+        This works by hiding the previous widget and showing the new
+        selection, creating the widget object if needed.
+
+        TODO: make sure the widget we set visible is fully updated with
+        the latest config file information.
+
+        Parameters
+        ----------
+        item : AtefItem
+            The selected item in the tree. This contains information like
+            the textual annotation, cached widget references, and
+            arguments for creating a new widget if needed.
+        """
+        if item is self.last_selection:
+            return
+        if self.last_selection is not None:
+            self.last_selection.get_widget().setVisible(False)
+        widget = item.get_widget()
+        if widget not in self.built_widgets:
+            self.layout().addWidget(widget)
+            self.built_widgets.add(widget)
+        widget.setVisible(True)
+        self.last_selection = item
+
+
+class AtefItem(QTreeWidgetItem):
+    """
+    A QTreeWidget item with some convenience methods.
+
+    Facilitates the widget creation/caching mechanisms.
+    """
+    widget_class: type[QWidget]
```

Comment:
I think it's appropriate to target newer python versions in user-facing application layers

Reply:
On refactor, I notice that this file mixes annotation styles...

Reply:
I think it's not a big deal, so long as we're in agreement that 3.9+ is what we're targeting.

The only downside is that it could limit what projects we import atef from - that is, some of our collaboration-facing projects we have held back to 3.7+, to my recollection. 

Given the nature of the atef package and where it sits dependency-wise, I expect that won't be a problem.